packages/flutter/lib/src/animation/animation_controller.dart

```diff
@@ -512,69 +517,21 @@ class AnimationController extends Animation<double>
   /// During the animation, [status] is reported as [AnimationStatus.reverse],
   /// which switches to [AnimationStatus.dismissed] when [lowerBound] is
   /// reached at the end of the animation.
-  TickerFuture reverse({ double? from }) {
-    assert(() {
-      if (duration == null && reverseDuration == null) {
-        throw FlutterError(
-          'AnimationController.reverse() called with no default duration or reverseDuration.\n'
-          'The "duration" or "reverseDuration" property should be set, either in the constructor or later, before '
-          'calling the reverse() function.',
-        );
-      }
-      return true;
-    }());
-    assert(
-      _ticker != null,
-      'AnimationController.reverse() called after AnimationController.dispose()\n'
-      'AnimationController methods should not be used after calling dispose.',
-    );
-    _direction = _AnimationDirection.reverse;
-    if (from != null) {
-      value = from;
-    }
-    return _animateToInternal(lowerBound);
-  }
+  TickerFuture reverse({ double? from }) => _toggle('reverse', false, from);
 
-  /// Toggles the direction of this animation, based on whether it [isForwardOrCompleted].
+  /// Toggles the direction of this animation.
   ///
-  /// Specifically, this function acts the same way as [reverse] if the [status] is
-  /// either [AnimationStatus.forward] or [AnimationStatus.completed], and acts as
-  /// [forward] for [AnimationStatus.reverse] or [AnimationStatus.dismissed].
-  ///
-  /// If [from] is non-null, it will be set as the current [value] before running
-  /// the animation.
+  /// If [toggleForward] is non-null, it determines the animation direction.
```

Comment:
Does this mean that `controller.toggle(true)` is the same as calling `controller.forward()` and `controller.toggle(false)` is the same as `controller.reverse()`?

From an API perspective I find it difficult to wrap my head around this API. It is not super-clear what the boolean means. Why is true == forward?

Reply:
> Does this mean that `controller.toggle(true)` is the same as calling `controller.forward()` and `controller.toggle(false)` is the same as `controller.reverse()`?

Yes!

Technically, we could get rid of `.forward()` and `.reverse()` and have everything implemented in terms of `.animateTo()` and `.animateFrom()` with relatively minimal refactoring. But I like having "forward" and "reverse" since they're common use cases.

Likewise, I think it makes sense to have `.toggle()`, since the following `AnimationController` is very common:

```dart
if (condition) {
  _controller.forward();
} else {
  _controller.reverse();
}
```

> From an API perspective I find it difficult to wrap my head around this API. It is not super-clear what the boolean means. Why is true == forward?

This is something I hadn't considered. After looking it over again, it does feel redundant to include the method in the parameter name, and I agree that it should have a better explanation.

I've made updates to the method accordingly; thank you!

Reply:
My problem with the API is more that when I see `controller.toggle(true)` I don't actually immediately know what that means. Toggle doesn't naturally take a boolean, so it makes the code harder to understand compared to the if/else pattern, which is immediately obvious.

Reply:
Ah okay, that makes sense: fewer lines of code isn't necessarily worth it if it becomes significantly more difficult to understand.

I 100% agree that `toggle(true)` and `toggle(false)` is extremely unintuitive—**animated_cross_fade.dart** is an egregious example of this:

```dart
_controller.toggle(switch (widget.crossFadeState) {
  CrossFadeState.showFirst => false,
  CrossFadeState.showSecond => true,
});


// much more readable as a switch statement
switch (widget.crossFadeState) {
  case CrossFadeState.showFirst:
    _controller.reverse();
  case CrossFadeState.showSecond:
    _controller.forward();
}
```

<br>

However, I believe that the vast majority of the `toggle` calls in this PR are very intuitive & easy to follow, for example:

```dart
@override
void didUpdateWidget(_RallyTab oldWidget) {
  super.didUpdateWidget(oldWidget);
  _controller.toggle(widget.isExpanded);
}
```

My immediate thought upon reading: "This looks like a stateful widget with an animation controller that toggles each time `isExpanded` changes."

<br>

If I were to give a letter grade to the current additions, it would probably be as follows:

**A+** for the empty `.toggle()` calls (6 in this PR)

<details> <summary><b>A-</b>: "simple" arguments (e.g. <code>widget.isExpanded</code>, 21 in this PR)</summary>

```dart
_controller.toggle(_showDrawerContents);

_controller.toggle(shouldOpenList);

_codeBackgroundColorController.toggle(
  _demoStateIndex.value == _DemoState.code.index,
);

_controller.toggle(widget.isExpanded);

_positionController.toggle(widget.value);

_controller.toggle(widget.mode == DatePickerMode.year);

enableController.toggle(widget.isEnabled);

selectController.toggle(widget.selected);

deleteDrawerController.toggle(hasDeleteButton);

_opacityController.toggle(widget.visible);

_orientationController.toggle(_orientationController.isDismissed);

_controller.toggle(widget.isExpanded);

_hoverColorController.toggle(widget.isHovering);

_controller.toggle(hasError);

_controller.toggle(widget.isSelected);

_extendedController.toggle(widget.extended);

enableController.toggle(isEnabled);

_state.enableController.toggle(isInteractive);

_controller.toggle(widget.isOpen);

_reactionFocusFadeController.toggle(focused);

_reactionHoverFadeController.toggle(hovering);
```

</details>

<details> <summary><b>B</b> for calls with a boolean or null-aware operator (7 examples)</summary>

```dart
_controller.toggle(widget.visibility?.value ?? true);

_controller.toggle(widget.visibility?.value ?? true);

_positionController.toggle(value ?? tristate);

avatarDrawerController.toggle(hasAvatar || widget.selected);

_controller.toggle(!widget.checked);

menuAnimation.toggle(!menuAnimation.isCompleted);

_floatingLabelController.toggle(
  _floatingLabelEnabled && widget._labelShouldWithdraw,
);
```

</details>

<details open> <summary><b>C</b> for the calls containing multiple operators (2 examples)</summary>

```dart
_state.overlayController.toggle(hovered && (hoveringStartThumb || hoveringEndThumb));

_moveController!.toggle(
  _moveController!.value > (widget.dismissThresholds[_dismissDirection] ?? _kDismissThreshold),
);
```

</details>

<details open> <summary><b>F</b> for animated_cross_fade.dart</summary>

```dart
_controller.toggle(switch (widget.crossFadeState) {
  CrossFadeState.showFirst => false,
  CrossFadeState.showSecond => true,
});
```

</details>

I'll go ahead and revert those 3 worst examples.